### PR TITLE
[OptionsResolver] support array of instance validation

### PR DESCRIPTION
**Pull request by EVODelavega**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15524 |
| License | MIT |
| Doc PR | symfony/symfony-docs#6048 |

A suggested approach to add support for typed-arrays as allowed types, using a simple recursive method.

Closing [PR 15671](https://github.com/symfony/symfony/pull/15671).

### CODE REVIEW

**Comment by EVODelavega** [comment]
PR was reviewed in previous PR, so setting status to reviewed now
Status: Reviewed

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 1]
```diff
@@ -896,6 +885,42 @@ public function offsetGet($option)
     }
 
     /**
+     * Verify value is of the allowed type. Recursive method to support
+     * typed array notation like ClassName[], or scalar arrays (int[]).
+     *
+     * @param string $type  the required allowedType string
+     * @param mixed  $value the value
+     *
+     * @return bool Whether or not $value if of the allowed type
```
Comment (xabbuh) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: is

Comment (jakzal) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: I'd say "Whether the $value is of the allowed type".

**Comment by stof** [comment]
there is a bunch of unrelated changes in this PR, making me think you messed the history a bit.

**Comment by nicolas-grekas** [comment]
Status: needs work (git history rebasing+squashing)

**Comment by Soullivaneuh** [comment]
@EVODelavega Are you also planning for a documentation PR about that?

**Comment by EVODelavega** [comment]
@Soullivaneuh Just looking at the comments. I'd be more than happy to create a documentation PR, after I've fixed the docblocks.

@nicolas-grekas: Do I really _have_ to rebase + squash everything? If so: squash everything into a single commit and rebase on master?

**Comment by EVODelavega** [comment]
@Soullivaneuh Created [documentation PR](https://github.com/symfony/symfony-docs/pull/6048)

File: CHANGELOG-3.0.md [hunk 2]
```diff
@@ -26,7 +26,7 @@ To get the diff between two versions, go to https://github.com/symfony/symfony/c
  * bug #16312 [HttpKernel] clearstatcache() so the Cache sees when a .lck file has been released (mpdude)
  * bug #16351 [WIP] [Form] [TwigBridge] Bootstrap horizontal theme missing tests (pieter2627)
  * feature #16715 [Form] Remove choices_as_values option on ChoiceType (nicolas-grekas)
- * feature #16692 [Form] Drop remaing CsrfProviderAdapter/Interface mentions (nicolas-grekas)
+ * feature #16692 [Form] Drop remaining CsrfProviderAdapter/Interface mentions (nicolas-grekas)
```
Comment (xabbuh) [inline CHANGELOG-3.0.md]: This change should be done in a different PR.

Comment (EVODelavega) [inline CHANGELOG-3.0.md]: Created a PR for the typo. Should I revert this change, or simply wait for the other PR to get merged and rebase the changes here?

**Comment by nicolas-grekas** [comment]
@EVODelavega you don't *have* to squash: we are able to do it while merging. But if you know how to do it, it would be great alongside with rewording the first commit message so that it makes more sense for some reader. But this is not mandatory :)
Status: needs review

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 3]
```diff
@@ -896,6 +885,42 @@ public function offsetGet($option)
...
+     * typed array notation like ClassName[], or scalar arrays (int[]).
+     *
+     * @param string $type  the required allowedType string
+     * @param mixed  $value the value
+     *
+     * @return bool Whether the $value is of the allowed type
+     */
+    private function verifyAllowedType($type, $value)
+    {
+        if (mb_substr($type, -2) === '[]') {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: all the calls to mb_substr should be replaced by plain substr calls imho

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 4]
```diff
@@ -896,6 +885,42 @@ public function offsetGet($option)
...
+            return true;
+        }
+
+        $type = isset(self::$typeAliases[$type]) ? self::$typeAliases[$type] : $type;
+
+        if (function_exists($isFunction = 'is_'.$type)) {
+            return $isFunction($value);
+        }
+
+        return ($value instanceof $type);
```
Comment (fabpot) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: The parenthesis can be removed.

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 5]
```diff
@@ -896,6 +885,42 @@ public function offsetGet($option)
...
+     *
+     * @param string $type  the required allowedType string
+     * @param mixed  $value the value
+     *
+     * @return bool Whether the $value is of the allowed type
+     */
+    private function verifyAllowedType($type, $value)
+    {
+        if (mb_substr($type, -2) === '[]') {
+            //allowed type is typed array
```
Comment (fabpot) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Can you add a space after `//` (same for the other comments)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 6]
```diff
@@ -963,16 +988,69 @@ public function count()
...
+            $vType = implode('|', $subTypes);
+        } else {
+            $vType = is_object($value) ? get_class($value) : gettype($value);
+        }
+
+        return sprintf(
+            '%s%s',
+            $vType,
+            $suffix
+        );
```
Comment (fabpot) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: should be on one line.

**Comment by fabpot** [comment]
This looks like a good new feature. What do you think @symfony/deciders?

File: src/Symfony/Component/OptionsResolver/CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+3.1.0
```
Comment (fabpot) [inline src/Symfony/Component/OptionsResolver/CHANGELOG.md]: Should be changed to 3.2.0 now.

**Comment by Tobion** [comment]
In general I think the feature is not bad. I'm just not sure if it's worth since there is an easy solution already: https://github.com/symfony/symfony/issues/15524#issuecomment-130369403

Also, the allowed types now look like phpdoc vars but do not actually support all of the phpdoc features like `int|float` or `(Foo|Bar)[]` are not possible. So it's kinda arbitrary what it now supports.

**Comment by fabpot** [comment]
I don't find the solution you mentioned that easy; and probably less than just having to add `[]`. I like the fact that we are reusing phpdoc notation here, and the fact that we do not support the whole notation yet does mean that this is arbitrary.

**Comment by xabbuh** [comment]
I am :+1: for supporting this. Looks quite reasonable and straight forward to me.

**Comment by Tobion** [comment]
Keep in mind the phpdoc interpretation of `int[]` implemented here does only support arrays of integers, not traversables of integers. This seems consistent with the current working state of PSR-5. For collections one would need to specifiy `\ArrayObject<string>` which is not implemented here.

**Comment by fabpot** [comment]
@Tobion I think this is not relevant for this PR as we are not implementing a phpdoc or PSR-5. We are just reusing something that people can understand easily as they are familiar with this notation.

**Comment by fabpot** [comment]
@EVODelavega Do you have time to take the comments into accounts?

**Comment by EVODelavega** [comment]
@fabpot It's been a while since I last looked at this PR, but I'm going to take all of the comments into account, work on it some more, and I'll let you know when it's ready.

Thanks.

**Comment by fabpot** [comment]
@EVODelavega Are you still interested in finishing this PR?

**Comment by bendavies** [comment]
Just came searching for this feature. Any interest in finishing this off? Or someone taking over?

**Comment by bendavies** [comment]
I don't mind taking over if OP has no time.

**Comment by Soullivaneuh** [comment]
> @EVODelavega Are you still interested in finishing this PR?

Or maybe someone else?

**Comment by pierredup** [comment]
@Soullivaneuh see #23112

**Comment by EVODelavega** [comment]
Ouch, is this still going? Erm, I suppose I can pick it up again should this be needed. It looks like the other PR implements this feature, though. I've switched from PHP to golang since opening this PR, so if there's an alternative PR that is more up-to-date, feel free to close this one.

**Comment by nicolas-grekas** [comment]
@EVODelavega would you be up rebasing this PR on 3.4 by the end of the week? Feature freeze is coming and there are some comments still pending also.
